### ISSUE TITLE
Single bit extract and add/sub could generate better code via carry bit

### ISSUE THREAD

**Issue by davezarzycki** [llvm/llvm-project#35256]
|  |  |
| --- | --- |
| Bugzilla Link | [35908](https://llvm.org/bz35908) |
| Version | trunk |
| OS | All |
| CC | @topperc,@davezarzycki,@LebedevRI,@RKSimon,@rotateright,@Tilka |

## Extended Description 
Consider the following code, which is a reduction derived from a client of llvm::TrailingObjects.

void dummy(int arg);
void test_add(int x, int y, int z) { dummy(bool(z & (1 << 30)) + x + y); }
void test_sub(int x, int y, int z) { dummy(bool(z & (1 << 30)) - x - y); }

On pre-BEXTR CPUs, the following code is emitted:

"test_add"
	shrl	$30, %edx
	andl	$1, %edx
	addl	%esi, %edi
	addl	%edx, %edi

"test_sub"
	shrl	$30, %edx
	andl	$1, %edx
	addl	%esi, %edi
	subl	%edi, %edx
	movl	%edx, %edi

And on BEXTR capable CPUs:

"test_add"
	movl	$286, %eax              # imm = 0x11E
	bextrl	%eax, %edx, %eax
	addl	%esi, %edi
	addl	%eax, %edi

"test_sub"
	movl	$286, %eax              # imm = 0x11E
	bextrl	%eax, %edx, %eax
	addl	%esi, %edi
	subl	%edi, %eax
	movl	%eax, %edi

But in practice, pre-BEXTR CPUs can generate the best code

"test_add"
	btl	$30, %edx
	adcl	%esi, %edx

"test_sub"
	btl	$30, %edx
	sbbl	%esi, %edx

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D122084

**Comment by RKSimon** [comment]
Fixed by 438ac282db97c584daf2d4d1a90e6b3d49ef9189

**Comment by davezarzycki** [comment]
Thanks you! :-)

**Comment by davezarzycki** [comment]
So I've been auditing the real world code gen of this (both clang and swift since they both use llvm::TrailingObjects). It seems like a refinement is possible. This is a common output pattern in clang, for example:

bt     $0x2a,%rdx
mov    $0x3,%esi
adc    $0xffffffff,%esi

Which could be optimized into:

xor    %esi,%esi
bt     $0x2a,%rdx
adc    $2,%esi

Am I missing something?

**Comment by RKSimon** [comment]
If we're not using the overflow output of adc/sbb we should be able to perform a number of optimizations like this - we had a ticket for missed X86ISD add/sub style combines but I can't find it right now.

**Comment by RKSimon** [comment]
@davezarzycki If can you send me a simplified repro I'll add the constant folded canonicalization

**Comment by davezarzycki** [comment]
This example is from llvm/lib/Target/X86/X86MCInstLower.cpp no less. :-)

If the following isn't enough context, I can get you the whole IR output:

```
define internal fastcc noundef i32 @_ZL7emitNopRN4llvm10MCStreamerEjPKNS_12X86SubtargetE(%"class.llvm::MCStreamer"* noundef nonnull align 8 dereferenceable(270) %0, i32 noundef %1, %"class.llvm::X86Subtarget"* noundef %2) unnamed_addr #0 {
  %4 = alloca %"class.llvm::MCInstBuilder", align 8
  %5 = alloca %"class.llvm::MCInstBuilder", align 8
  %6 = alloca %"class.llvm::MCInstBuilder", align 8
  %7 = getelementptr inbounds %"class.llvm::X86Subtarget", %"class.llvm::X86Subtarget"* %2, i64 0, i32 9
  %8 = load i8, i8* %7, align 4, !tbaa !379, !range !11
  %9 = icmp eq i8 %8, 0
  br i1 %9, label %22, label %10

10:                                               ; preds = %3
  %11 = getelementptr inbounds %"class.llvm::X86Subtarget", %"class.llvm::X86Subtarget"* %2, i64 0, i32 0, i32 0, i32 0, i32 13, i32 0, i64 1
  %12 = load i64, i64* %11, align 8, !tbaa !120
  %13 = and i64 %12, 35184372088832
  %14 = icmp eq i64 %13, 0
  br i1 %14, label %15, label %22

15:                                               ; preds = %10
  %16 = and i64 %12, 140737488355328
  %17 = icmp eq i64 %16, 0
  br i1 %17, label %18, label %22

18:                                               ; preds = %15
  %19 = and i64 %12, 70368744177664
  %20 = icmp eq i64 %19, 0
  %21 = select i1 %20, i32 10, i32 11
  br label %22

22:                                               ; preds = %18, %15, %10, %3
  %23 = phi i32 [ 1, %3 ], [ 7, %10 ], [ 15, %15 ], [ %21, %18 ]
  %24 = getelementptr inbounds %"class.llvm::X86Subtarget", %"class.llvm::X86Subtarget"* %2, i64 0, i32 7
  %25 = load i8, i8* %24, align 2, !tbaa !380, !range !11
  %26 = icmp eq i8 %25, 0
  %27 = select i1 %26, i32 %23, i32 2
  %28 = tail call i32 @llvm.umin.i32(i32 %27, i32 %1)
  %29 = add nsw i32 %28, -1
...
```

Generates the following assembly:

```
00000000016cb880 <_ZL7emitNopRN4llvm10MCStreamerEjPKNS_12X86SubtargetE>:
 16cb880:       push   %rbp
 16cb881:       push   %r15
 16cb883:       push   %r14
 16cb885:       push   %r13
 16cb887:       push   %r12
 16cb889:       push   %rbx
 16cb88a:       sub    $0xc8,%rsp
 16cb891:       mov    %rdx,%r14
 16cb894:       mov    %rdi,%rbx
 16cb897:       cmpb   $0x0,0x114(%rdx)
 16cb89e:       je     16cb8ce <_ZL7emitNopRN4llvm10MCStreamerEjPKNS_12X86SubtargetE+0x4e>
 16cb8a0:       mov    0xc8(%r14),%rax
 16cb8a7:       mov    $0x7,%ecx
 16cb8ac:       bt     $0x2d,%rax
 16cb8b1:       jb     16cb8d3 <_ZL7emitNopRN4llvm10MCStreamerEjPKNS_12X86SubtargetE+0x53>
 16cb8b3:       mov    $0xf,%ecx
 16cb8b8:       bt     $0x2f,%rax
 16cb8bd:       jb     16cb8d3 <_ZL7emitNopRN4llvm10MCStreamerEjPKNS_12X86SubtargetE+0x53>
 16cb8bf:       bt     $0x2e,%rax
 16cb8c4:       mov    $0xb,%ecx
 16cb8c9:       adc    $0xffffffff,%ecx
...
```

**Comment by RKSimon** [comment]
Thank you - it looks like its close enough to https://reviews.llvm.org/D122482 that it will get caught by the same fold.

**Comment by RKSimon** [comment]
> Thank you - it looks like its close enough to https://reviews.llvm.org/D122482 that it will get caught by the same fold.

This should be fixed by 6697e3354fbe51242729888529d047c646acdab3
